### ISSUE TITLE
quality-suite TIMEOUT: quality_vs_gamlss_binomial_location_scale — unidentifiable Bernoulli scale smooth drives exact joint spatial optimizer to hang

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#717]
## Summary

The reference-quality test `quality_vs_gamlss_binomial_location_scale` **times out** under the workflow's `timeout 360` wrapper (`rc=124`). The hang is in the **second** subtest, the real-data arm `gam_binomial_location_scale_logit_p_matches_mgcv_binomial_on_real_data`, during gam's **own** native fit — not in any mgcv reference call. The native fit feeds a statistically **unidentifiable** Bernoulli (n=1) location-scale spec — a free *smooth* `log σ(pc1, pc2)` axis — into the **exact two-block spatial joint optimizer**, which then searches a flat likelihood ridge (zero curvature along the `t`/`σ` trade-off) and never certifies convergence.

## Timeout evidence

- The quality-suite wrapper killed the process at the **360 s** cap → `rc=124`.
- The **first** subtest, `gam_binomial_location_scale_logit_p_matches_mgcv_binomial` (`tests/quality_vs_gamlss_binomial_location_scale.rs:80`), had already **panicked** with:

  > `log-sigma smooth collapsed to a constant (range ...); the joint scale axis did not fit`

  (assertion at `tests/quality_vs_gamlss_binomial_location_scale.rs:273-275`). That panic is itself a symptom of the same non-identifiability — the scale axis carries no signal — but the process keeps running into the next subtest.
- The captured log then stops **right after the real-data subtest starts** (`gam_binomial_location_scale_logit_p_matches_mgcv_binomial_on_real_data`, `tests/quality_vs_gamlss_binomial_location_scale.rs:385`) and **before any mgcv reference call**. So the time is being burned inside gam's native fit, not in R/mgcv.

## Where the time goes

The real-data native fit at `tests/quality_vs_gamlss_binomial_location_scale.rs:451-458`:

```rust
let cfg = FitConfig {
    family: Some("binomial".to_string()),
    noise_formula: Some("1 + s(pc1, bs='tp') + s(pc2, bs='tp')".to_string()),
    ..FitConfig::default()
};
let result = fit_from_formula("y ~ s(pc1, bs='tp') + s(pc2, bs='tp')", &train_ds, &cfg)
    .expect("gam binomial location-scale fit on prostate train");
```

This is a two-block GAMLSS: a smooth latent threshold `t(pc1, pc2)` (mean formula) **and** a smooth `log σ(pc1, pc2)` (`noise_formula`), both `bs='tp'` thin-plate spatial terms with their own length-scale / smoothing hyperparameters.

`BinomialLocationScaleTermBuilder` forces the **exact** spatial joint optimizer:

```rust
fn exact_spatial_joint_supported(&self) -> bool { true }   // src/families/gamlss.rs:3256-3258
fn require_exact_spatial_joint(&self) -> bool { true }      // src/families/gamlss.rs:3260-3262
```

With `require_exact_spatial_joint() == true`, there is no cheaper fallback path: the solver must search the full **threshold-smoothing × scale-smoothing × spatial-length-scale** product of hyperparameter directions for both `tp` blocks. Several of those directions (everything along the scale axis) have essentially no likelihood curvature, so the outer REML/length-scale search keeps probing without ever satisfying its convergence criterion until the 360 s wall-clock cap kills it.

## Root cause

This is **statistical non-identifiability driving a solver pathology**, not merely a large dataset.

For 0/1 Bernoulli data the model's link-scale predictor is the composite

$$ q \;=\; -\,t/\sigma \;=\; -\,t\,\exp(-\eta_{ls}) \;=\; \operatorname{logit} P(y=1), \qquad P(y=1) = \operatorname{expit}(q). $$

The per-observation log-likelihood for `y_i \in \{0,1\}` is

$$ \ell_i(t,\sigma) \;=\; y_i \log \operatorname{expit}(q_i) + (1-y_i)\log\bigl(1-\operatorname{expit}(q_i)\bigr), $$

which depends on the parameters **only through the scalar** `q_i = -t_i/\sigma_i`. Hence the likelihood is invariant under the one-parameter rescaling

$$ t \to c\,t, \quad \sigma \to c\,\sigma \quad (c>0) \;\Longrightarrow\; q \text{ unchanged}, $$

so `t` and `σ` are **not separately identifiable** — only their ratio `q` is. This is documented in the test itself at `tests/quality_vs_gamlss_binomial_location_scale.rs:37-45` ("only the COMPOSITE `q = -t/σ` (= logit P) is identifiable from 0/1 data … any rescaling `t → c·t, σ → c·σ` leaves the likelihood unchanged").

Concretely, the Fisher information is **rank-deficient** along the trade-off direction. Writing `w_i = \operatorname{expit}(q_i)\,(1-\operatorname{expit}(q_i)) > 0`, the score and information decompose through the single channel `q`:

$$ \frac{\partial \ell_i}{\partial \theta} = (y_i - \operatorname{expit}(q_i))\,\frac{\partial q_i}{\partial \theta}, \qquad \mathcal{I}(\theta) = \sum_i w_i\,\frac{\partial q_i}{\partial \theta}\frac{\partial q_i}{\partial \theta}^{\!\top}. $$

Because `q_i = -t_i\exp(-\eta_{ls,i})`, the gradient w.r.t. the threshold direction and the scale direction are **collinear** at the optimum:

$$ \frac{\partial q_i}{\partial t_i} = -\exp(-\eta_{ls,i}) = \frac{q_i}{t_i}, \qquad \frac{\partial q_i}{\partial \eta_{ls,i}} = t_i\exp(-\eta_{ls,i}) = -q_i, $$

so `\partial q_i/\partial \eta_{ls,i} = -t_i \,(\partial q_i/\partial t_i)`. The information matrix is therefore singular along the `(t, \sigma)` ridge: moving `t` and `\eta_{ls}` together so that `q` is held fixed leaves `\ell` **exactly flat**. The penalty pins the overall magnitude weakly, but there is **no data curvature** to localize the split between `t` and `\sigma`.

The test nevertheless demands a **nonconstant** smooth `log σ(pc1, pc2)` (it even asserts `scale_block.beta.len() >= 2` at `tests/quality_vs_gamlss_binomial_location_scale.rs:473-477` and a nonzero range at lines 273-275 in the first subtest). Because `bs='tp'` makes the scale block a thin-plate spatial term with its own length-scale hyperparameter, the **exact** joint optimizer ends up searching a multi-dimensional smoothing-parameter space in which the scale block contributes a near-flat (rank-deficient) direction. With no curvature to certify a minimum, the outer search iterates indefinitely → the 360 s timeout.

## Suggested fix

Do **not** fit a separate free smooth `noise_formula` for Bernoulli n=1 data:

- Add an **early identifiability guard** that rejects a binomial (n=1) location-scale spec carrying a *free smooth* scale axis, before it ever reaches `require_exact_spatial_joint`. For n=1 Bernoulli the scale `σ` is unidentified, so a smooth `log σ(x)` is ill-posed by construction.
- Concretely, for Bernoulli n=1 either **freeze σ** (constant, e.g. `σ ≡ 1`, fitting only the identifiable threshold `t = -q`), or **drop the scale smooth** entirely and fit/assert only the identifiable composite `P = \operatorname{expit}(-t/\sigma)`.
- If a location-scale binomial path is to be kept at all, restrict the free smooth scale axis to **count** binomial data (`n > 1`, the `DBI`-style variance-inflation regime) where σ is identifiable, and reject it on n=1.
- The guard should fail **loudly and immediately** (a typed error from the family builder) rather than letting the exact joint optimizer spin.

The test would then assert only the identifiable `\operatorname{expit}(-t/\sigma)` truth-recovery metric and drop the non-degeneracy assertion on the unidentified scale axis.

## Classification

**Genuine defect.** A missing identifiability guard lets an ill-posed Bernoulli (n=1) location-scale specification — a free *smooth* scale axis whose Fisher information is rank-deficient along the `t`/`σ` trade-off — into the exact two-block spatial joint optimizer (`require_exact_spatial_joint() -> true`, `src/families/gamlss.rs:3260-3262`). The optimizer then searches a flat likelihood ridge and never converges, hanging until the suite's 360 s cap.

## Related issues

- #660 — GAMLSS location-scale identifiability audit.
- #566, #501 — location-scale fitting / calibration.
- #684, #685, #686, #687, #688 — location-scale calibration cluster.
- #406, #430, #469 — related location-scale / scale-axis work.

## Reproduction

```
cargo test --test quality_vs_gamlss_binomial_location_scale
```

(Do not run interactively — the second subtest hangs to the 360 s timeout; `rc=124` under the suite's `timeout` wrapper.)

_Filed from the reference-quality-suite TIMEOUT review._

**Comment by SauersML** [comment]
## Deeper source-level mechanism (follow-up)

The original report correctly identifies non-identifiability and pins the *hang* on the exact two-block spatial joint optimizer. This follow-up traces the precise **upstream** root cause in the family code and shows that the spatial joint optimizer is merely where a pre-existing flat objective gets ground out — the actual defect is an **unfixed nonlinear scale/location gauge** that the current "identification" step provably cannot remove. All code below was read directly from `src/families/gamlss.rs` and `src/families/scale_design.rs`.

### 1. The likelihood sees only the composite index `q = q0 + ηwiggle`

For binomial/Bernoulli location-scale the link-scale index is built from the single function

```rust
fn binomial_location_scale_q0(eta_t: f64, sigma: f64) -> f64 {
    -eta_t / sigma
}
```
(`src/families/gamlss.rs:5215-5217`), with `σ = exp(η_ls)`. The per-observation log-likelihood consumes only `q`:

```rust
InverseLink::Standard(StandardLink::Logit) => Ok(weight
    * (-y * crate::linalg::utils::stable_softplus(-q)
        - (1.0_f64 - y) * crate::linalg::utils::stable_softplus(q))),
```
(`binomial_location_scale_log_likelihood`, `src/families/gamlss.rs:5234-5236`).

Because the likelihood depends on `(η_t, σ)` **only** through the scalar `q = −η_t/σ`, the threshold field `η_t(x)` and the scale field `σ(x)` are **not separately identifiable**. For *any* positive function `a(x)`, the pair `(η_t', σ') = (a·η_t, a·σ)` yields identical `q(x)` and therefore identical likelihood. This is a **multiplicative, nonlinear (ratio) gauge** — not an additive one.

### 2. The "identification" step is purely *linear column residualization* — it cannot touch a nonlinear ratio gauge

The repair the code applies is `identified_binomial_log_sigma_design` (`src/families/gamlss.rs:1831-1851`), invoked at the binomial location-scale build sites (e.g. `src/families/gamlss.rs:3282`, `:3337`, `:3441`, `:3527`). It calls `build_scale_deviation_transform_design` and then `build_scale_deviation_operator`. The entire effect of that operator is this chunk transform:

```rust
// Active columns: residual = noise - projection, then center & rescale.
if first_active < p_noise {
    let fitted =
        apply_projection_chunk(primary_chunk, &transform.projection_coef, first_active);
    for j in first_active..p_noise {
        let jj = j - first_active;
        let scale = transform.rescale[j];
        let center = transform.weighted_column_mean[j];
        for i in 0..rows {
            out[[i, j]] = (noise_chunk[[i, j]] - fitted[[i, jj]] - center) * scale;
        }
    }
}
```
(`apply_scale_deviation_reparam_chunk`, `src/families/scale_design.rs:831-843`).

This residualizes each **scale design column** against a *linear* projection onto the threshold columns, then centers and rescales. That removes **additive linear overlap** between the `η_t` and `η_ls` column spaces. It does **nothing** to the gauge that actually plagues Bernoulli data, which lives in the *ratio* `η_t/σ` and is **nonlinear** in `η_ls` (via `σ = exp(η_ls)`). No linear residualization of columns can quotient out a `(η_t, σ) → (a·η_t, a·σ)` rescaling symmetry. The gauge survives the "identification" intact.

### 3. The only thing that pins the scale block is a full identity shrinkage penalty

After the (ineffective-for-this-gauge) reparam, the code appends a dense identity penalty to the entire log-σ block:

```rust
let mut log_sigma_penalty_matrices: Vec<PenaltyMatrix> =
    noise_design.penalties_as_penalty_matrix();
log_sigma_penalty_matrices.push(PenaltyMatrix::Dense(identity_penalty(p_noise)));
```
(`src/families/gamlss.rs:3284-3286`; identical at `:3443-3445`; `identity_penalty` at `:1853-1859`; standalone shrinkage appender `append_binomial_log_sigma_shrinkage_penalty_design` at `:1861`). `identity_penalty` is just `I`:

```rust
fn identity_penalty(dim: usize) -> Array2<f64> {
    let mut penalty = Array2::<f64>::zeros((dim, dim));
    for i in 0..dim {
        penalty[[i, i]] = 1.0;
    }
    penalty
}
```

So along the scale axis the *only* curvature in the penalized objective comes from this ridge penalty — the likelihood contributes none (it is flat along the gauge direction by §1).

### 4. This single root explains BOTH symptoms

- **Synthetic arm (first subtest panic: "log-sigma smooth collapsed to a constant").** After the gauge, the scale field carries no likelihood information; the identity shrinkage penalty (§3) is the only force acting on it, so REML drives the log-σ coefficients toward zero and the smooth collapses to a constant — exactly the asserted failure.
- **Real-data arm (360 s timeout).** The custom-family two-block optimizer is profiling an objective that is flat/near-flat along an entire (gauge) direction of one block. With no data curvature to localize the `η_t`/`σ` split, the outer REML/length-scale search never certifies convergence and spins to the wall-clock cap.

### 5. Relationship to the original framing

The original issue attributes the hang to the **exact spatial joint optimizer** (`require_exact_spatial_joint() -> true`, `src/families/gamlss.rs:3260-3262`). That is accurate as the *locus* of the hang, but it is downstream. The exact joint optimizer is simply where a flat objective gets ground out; it would converge fine on an identifiable problem. The **upstream cause** is that the nonlinear scale/location ratio gauge of §1 is never actually removed — the linear column residualization of §2 is the wrong tool for a nonlinear (multiplicative) symmetry, and the identity penalty of §3 papers over the resulting flatness rather than resolving it.

### 6. Proposed fix (sharpened)

- **Do not fit a free smooth `log σ(x)` for Bernoulli n=1 data.** It is ill-posed by construction.
- **Fix the nonlinear gauge at its source** rather than via linear column residualization: anchor σ (e.g. `σ ≡ 1`, fitting only the identifiable threshold `t = −q`), or impose an explicit nonlinear constraint on the ratio — not a column projection.
- Alternatively, **assert/fit only the identifiable composite** `P = expit(−t/σ)` and drop the non-degeneracy assertion on the unidentified scale axis.
- Add an **early identifiability guard** in the binomial location-scale builder that loudly rejects a free *smooth* scale axis on n=1 data **before** it reaches `require_exact_spatial_joint`, so the optimizer never spins on a gauge direction.

_Follow-up to the reference-quality-suite TIMEOUT review; source-verified._

**Comment by SauersML** [comment]
Verified fixed on MSI compute (free node): `quality_vs_gamlss_binomial_location_scale` passes (2/2 subtests, 0.2s). The principled identifiability guard rejects the unidentifiable free smooth log-σ for binomial location-scale before the exact-joint optimizer is entered, so the former 360s hang is gone. Closed after direct test execution.
